README
Static Gallery
==============

Static Web Gallery Generator using Jinja2

Template selection

TODO
====

- Select template by command line
- Force rebuild gallery (or not)
- Update the gallery instead of creating again entirely


CHANGES.txt
v0.1.0, 2013-02-28 -- Initial release.


app.py
"""
Small web service for static gallery generation.

"""

import flask
from flask import Flask
from flask import render_template
from PIL import Image
import os

app = Flask(__name__)
app.config.from_pyfile("configuration.py")

#FIXME Provisional
static_name = app.config['STATIC_NAME']
gallery_name = app.config['GALLERY_NAME']
thumbs_name = app.config['THUMBS_NAME']
index_gallery_name = app.config['INDEX_GALLERY_NAME']
secret_prefix_name = app.config['SECRET_PREFIX_NAME']

# url parse is ugly
static_url = '/%s/'
gallery_base_url = '/%s/%s/%s/'
thumbs_base_url = '/%s/%s/%s/%s/'

static_base_path = '%s/%s/'
gallery_base_path = '%s/%s/%s/'
thumbs_base_path = '%s/%s/%s/%s/'

thumbs_size = (180, 120)
image_format = "JPEG"

class Item(object):
	def __init__(self, title, url):
		self.title = title
		self.url = url

def get_gallery_urls(folder):
	"""
	Get gallery and thumbs url

	"""

	gallery_url = gallery_base_url % (static_name, gallery_name, folder)
	thumbs_url = thumbs_base_url % (static_name, gallery_name, folder, thumbs_name)

	return gallery_url, thumbs_url

def get_gallery_path(folder):
	"""
	Get gallery and thumbs path

	"""

	gallery_path = os.path.join(app.config['ACTUAL_PATH'], gallery_base_path % (static_name, gallery_name, folder))
	thumbs_path = os.path.join(app.config['ACTUAL_PATH'], thumbs_base_path % (static_name, gallery_name, folder, thumbs_name))

	return gallery_path, thumbs_path

@app.route('/')
def do_menu():
	"""
	Generate dynamic menu

	"""

	index_image_url = (gallery_base_url % (static_name, gallery_name, app.config['INDEX_IMAGE_NAME']))[:-1]
	static_path =os.path.normcase(os.path.join(app.config['ACTUAL_PATH'], static_base_path % (static_name, gallery_name)))
	galleries = os.listdir(static_path)

	items = []

	for gallery in galleries:
		# check each gallery
		if gallery.startswith(secret_prefix_name) == False and gallery != app.config['INDEX_IMAGE_NAME']:
			# do not display secret one
			gallery_url, thumbs_url = get_gallery_urls(gallery)
			items.append(Item(gallery, gallery_url))

	return render_template('menu.jinja2', title=app.config['INDEX_TITLE'], items=items, image_url=index_image_url)

@app.route('/do_gallery/')
def do_gallery():
    for elem in os.listdir(os.path.join(app.config['ACTUAL_PATH'], static_name, gallery_name)):
        if os.path.isdir(os.path.join(app.config['ACTUAL_PATH'], static_name, gallery_name, elem)):
            gallery_path, thumbs_path = get_gallery_path(elem)
            image_list = os.listdir(gallery_path)

            if thumbs_name in image_list:
                # suppress thumbnail folder
                image_list.remove(thumbs_name)

            if index_gallery_name in image_list:
                # simply remove
                image_list.remove(index_gallery_name)

            if os.path.exists(thumbs_path) == False:
                # create thumbs directory
                os.mkdir(thumbs_path)

            image_list.sort()

            for image in image_list:
                # create thumbs
                image_path = os.path.join(gallery_path, image)
                image_thumb_path = os.path.join(thumbs_path, image)

                image_full = Image.open(os.path.normcase(image_path))
                image_full.thumbnail(thumbs_size, Image.ANTIALIAS)
                image_full.save(image_thumb_path, image_full.format)

            # generate template
            gallery_url, thumbs_url = get_gallery_urls(elem)
            template = render_template('base.jinja2', title=elem, thumbs_url=thumbs_url, gallery_url=gallery_url, image_list=image_list)

            # save static file
            new_file = file(os.path.join(gallery_path, index_gallery_name), 'w')
            new_file.write(template)
            new_file.close()

    # FIXME Provisional
    gallery_url = "index.html"

    return flask.redirect(gallery_url + index_gallery_name, code=302)

if __name__ == '__main__':
    app.run(host='0.0.0.0', port=5000)



static_gallery/utils.py
import os
import SimpleHTTPServer
import SocketServer
from PIL import Image
import ExifTags
from jinja2 import Environment, PackageLoader

from configuration import *


env = Environment(loader=PackageLoader('static_gallery',
                  'templates'))


def generate_html_output(path, template_name, arguments):
    template = env.get_template(template_name)
    template_content = template.render(**arguments)

    # save static file
    new_file = file(path, 'w')
    new_file.write(template_content)
    new_file.close()


def create_menu(dst_path, template_name):
        galleries = []
        gallery_path = dst_path

        for gallery in os.listdir(gallery_path):
            gallery_elem_path = os.path.join(gallery_path, gallery)

            if os.path.isdir(gallery_elem_path) and gallery != "static":
                galleries.append({"title": gallery,
                                  "url": "/".join(["/" + gallery, "index.html"])})

        generate_html_output("/".join([gallery_path, "index.html"]),
                             template_name,
                             {"title": "Album", "galleries": galleries})


def prepare_images(src_path, dst_path, gallery_name, image_list, reload_gallery):
    src_gallery = os.path.join(src_path, gallery_name)
    dst_gallery = os.path.join(dst_path, gallery_name)
    thumbs_path = os.path.join(dst_gallery, THUMBS_NAME)

    if not os.path.exists(dst_gallery):
        os.mkdir(dst_gallery)

    if not os.path.exists(thumbs_path):
        os.mkdir(thumbs_path)

    for image in image_list:
        # create thumbs
        image_path = os.path.join(src_gallery, image)
        image_thumb_path = os.path.join(thumbs_path, image)

        image_full = Image.open(os.path.normcase(image_path))
        image_exif_dict = image_full._getexif()

        if image_exif_dict is not None:
            exif = {}

            for k, v in image_exif_dict.items():
                try:
                    if ExifTags.TAGS[k] == 'Orientation':
                        exif[ExifTags.TAGS[k]] = v
                        break
                except KeyError:
                    print("the image has no metadata")

            if exif['Orientation'] == 3:
                image_full = image_full.rotate(180, expand=True)
            elif exif['Orientation'] == 6:
                image_full = image_full.rotate(270, expand=True)
            elif exif['Orientation'] == 8:
                image_full = image_full.rotate(90, expand=True)

        image_dst_path = os.path.join(dst_gallery, image)
        # TODO An option to rewrite images
        if not os.path.exists(image_dst_path) or reload_gallery:
            # FIXME PIL overrides original metadata when saving, we want to keep it
            image_full.save(image_dst_path, image_full.format)

        if not os.path.exists(os.path.join(image_thumb_path)) or reload_gallery:
            image_full.thumbnail(THUMBS_SIZE, Image.ANTIALIAS)
            image_full.save(image_thumb_path, image_full.format)


def create_gallery(src_path, dst_path, template_name, reload_gallery):
    gallery_path = src_path

    for gallery_name in os.listdir(gallery_path):
        gallery_elem_path = os.path.join(gallery_path, gallery_name)
        dst_gallery_path = os.path.join(dst_path, gallery_name)

        if os.path.isdir(gallery_elem_path):
            image_list = os.listdir(gallery_elem_path)
            image_list.sort()
            prepare_images(src_path, dst_path, gallery_name, image_list, reload_gallery)

            gallery_url = "".join(["/",  gallery_name, "/"])
            thumbs_url = "".join([gallery_url, THUMBS_NAME, "/"])

            generate_html_output("/".join([dst_gallery_path, "index.html"]),
                                 template_name,
                                 {"title": gallery_name.capitalize(),
                                  "thumbs_url": thumbs_url,
                                  "gallery_url": gallery_url,
                                  "image_list": image_list})


def exec_server(directory=DST_GALLERY_PATH, port=8000):
    """ Executes a server for development or simple gallery viewing """

    os.chdir(directory)
    print("Changing to the gallery directory... \n%s" % directory)
    handler = SimpleHTTPServer.SimpleHTTPRequestHandler
    httpd = SocketServer.TCPServer(("", port), handler)

    print("Serving at port %d" % port)
    httpd.serve_forever()


def process_call(arguments):
    """ Process call arguments """

    src_path = arguments.src or SRC_GALLERY_PATH
    dst_path = arguments.dst or DST_GALLERY_PATH
    template_gallery = arguments.template_gallery or "galleria.jinja2"
    template_menu = arguments.template_menu or "menu.jinja2"
    reload_gallery = arguments.reload

    if arguments.port is not None:
        port = int(arguments.port)
    else:
        port = 8000

    if not os.path.exists(src_path):
        print 'The source path does not exist: %s' % src_path
        return False
    if not os.path.exists(dst_path):
        print 'The destiny path does not exist: %s' % dst_path
        return False

    create_gallery(src_path, dst_path, template_gallery, reload_gallery)
    create_menu(dst_path, template_menu)

    if arguments.server:
        exec_server(port=port)


static_gallery/generator.py
"""
    Static Gallery Generator

"""

from PIL import Image
import ExifTags
import os
from jinja2 import Environment, PackageLoader
from configuration import *
import sys
import argparse
import SimpleHTTPServer
import SocketServer


sys.path.append(os.path.join(os.path.dirname(__file__), os.path.pardir))
env = Environment(loader=PackageLoader('static_gallery_generator',
                  'templates'))


def generate_html_output(path, template_name, arguments):
    template = env.get_template(template_name)
    template_content = template.render(**arguments)

    # save static file
    new_file = file(path, 'w')
    new_file.write(template_content)
    new_file.close()


def create_menu(dst_path, template_name):
        galleries = []
        gallery_path = dst_path

        for gallery in os.listdir(gallery_path):
            gallery_elem_path = os.path.join(gallery_path, gallery)

            if os.path.isdir(gallery_elem_path) and gallery != "static":
                galleries.append({"title": gallery,
                                  "url": "/".join(["/" + gallery, "index.html"])})

        generate_html_output("/".join([gallery_path, "index.html"]),
                             template_name,
                             {"title": "Album", "galleries": galleries})


def prepare_images(src_path, dst_path, gallery_name, image_list, reload_gallery):
    src_gallery = os.path.join(src_path, gallery_name)
    dst_gallery = os.path.join(dst_path, gallery_name)
    thumbs_path = os.path.join(dst_gallery, THUMBS_NAME)

    if not os.path.exists(dst_gallery):
        os.mkdir(dst_gallery)

    if not os.path.exists(thumbs_path):
        os.mkdir(thumbs_path)

    for image in image_list:
        # create thumbs
        image_path = os.path.join(src_gallery, image)
        image_thumb_path = os.path.join(thumbs_path, image)

        image_full = Image.open(os.path.normcase(image_path))
        image_exif_dict = image_full._getexif()

        if image_exif_dict is not None:
            exif = {}

            for k, v in image_exif_dict.items():
                try:
                    if ExifTags.TAGS[k] == 'Orientation':
                        exif[ExifTags.TAGS[k]] = v
                        break
                except KeyError:
                    print("the image has no metadata")

            if exif['Orientation'] == 3:
                image_full = image_full.rotate(180, expand=True)
            elif exif['Orientation'] == 6:
                image_full = image_full.rotate(270, expand=True)
            elif exif['Orientation'] == 8:
                image_full = image_full.rotate(90, expand=True)

        image_dst_path = os.path.join(dst_gallery, image)
        # TODO An option to rewrite images
        if not os.path.exists(image_dst_path) or reload_gallery:
            # FIXME PIL overrides original metadata when saving, we want to keep it
            image_full.save(image_dst_path, image_full.format)

        if not os.path.exists(os.path.join(image_thumb_path)) or reload_gallery:
            image_full.thumbnail(THUMBS_SIZE, Image.ANTIALIAS)
            image_full.save(image_thumb_path, image_full.format)


def create_gallery(src_path, dst_path, template_name, reload_gallery):
    gallery_path = src_path

    for gallery_name in os.listdir(gallery_path):
        gallery_elem_path = os.path.join(gallery_path, gallery_name)
        dst_gallery_path = os.path.join(dst_path, gallery_name)

        if os.path.isdir(gallery_elem_path):
            image_list = os.listdir(gallery_elem_path)
            image_list.sort()
            prepare_images(src_path, dst_path, gallery_name, image_list, reload_gallery)

            gallery_url = "".join(["/",  gallery_name, "/"])
            thumbs_url = "".join([gallery_url, THUMBS_NAME, "/"])

            generate_html_output("/".join([dst_gallery_path, "index.html"]),
                                 template_name,
                                 {"title": gallery_name.capitalize(),
                                  "thumbs_url": thumbs_url,
                                  "gallery_url": gallery_url,
                                  "image_list": image_list})


def exec_server(directory=DST_GALLERY_PATH, port=8000):
    """ Executes a server for development or simple gallery viewing """

    os.chdir(directory)
    print("Changing to the gallery directory... \n%s" % directory)
    handler = SimpleHTTPServer.SimpleHTTPRequestHandler
    httpd = SocketServer.TCPServer(("", port), handler)

    print("Serving at port %d" % port)
    httpd.serve_forever()


def process_call(arguments):
    """ Process call arguments """

    src_path = arguments.src or SRC_GALLERY_PATH
    dst_path = arguments.dst or DST_GALLERY_PATH
    template_gallery = arguments.template_gallery or "galleria.jinja2"
    template_menu = arguments.template_menu or "menu.jinja2"
    reload_gallery = arguments.reload

    if arguments.port is not None:
        port = int(arguments.port)
    else:
        port = 8000

    create_gallery(src_path, dst_path, template_gallery, reload_gallery)
    create_menu(dst_path, template_menu)

    if arguments.server:
        exec_server(port=port)


def main():
    """Parses app command line options """
    parser = argparse.ArgumentParser(description='Static Gallery Generator Options.')

    parser.add_argument('--template-gallery',
                        help='Choose the name of the template for the gallery')
    parser.add_argument('--template-menu',
                        help='Choose the template for the menu')
    parser.add_argument('--src',
                        help='Source directory for the gallery')
    parser.add_argument('--dst',
                        help='Destiny where the web gallery will be generated')
    parser.add_argument('--server', action='store_true',
                        help='Executes a server')
    parser.add_argument('--port',
                        help='Choose the port for the server, by default 8000')
    parser.add_argument('--reload', action='store_true',
                        help="Reload all galleries, even if they exist in the destiny")

    arguments = parser.parse_args()
    process_call(arguments)


if __name__ == '__main__':
    main()
